asf_heat_pump_suitability/pipeline/prepare_features/grid_capacity.py

```diff
@@ -0,0 +1,758 @@
+
+    Returns:
+        DataFrame with distributed headroom per LSOA.
+    """
+    # Merge LSOA boundaries with household counts
+    lsoa_with_households = lsoa_gdf.merge(households_df, on="LSOA21CD", how="left")
+
+    # Perform spatial join between LSOAs and substations
+    joined = gpd.sjoin(
+        lsoa_with_households, substations_df, how="inner", predicate="intersects"
+    )
+
+    # Calculate total households served by each substation
+    substation_total_households = (
+        joined.groupby("index_right")["household_count"].sum().reset_index()
+    )
+    substation_total_households = substation_total_households.rename(
+        columns={"household_count": "total_households"}
+    )
+
+    # Merge total households back to joined dataframe
+    joined = joined.merge(substation_total_households, on="index_right")
+
+    # Calculate the fraction of substation's load for each LSOA
+    joined["load_fraction"] = joined["household_count"] / joined["total_households"]
+
+    # Distribute headroom based on load fraction
+    joined["distributed_headroom"] = joined["headroom_mva"] * joined["load_fraction"]
+
+    # Aggregate distributed headroom by LSOA
+    lsoa_headroom = (
+        joined.groupby("LSOA21CD")["distributed_headroom"].sum().reset_index()
+    )
+
+    return lsoa_headroom
+
+
+def calculate_headroom_per_household(
+    lsoa_headroom: pd.DataFrame, households_df: pd.DataFrame
+) -> pd.DataFrame:
+    """
+    Calculate headroom per household for each LSOA.
+
+    Args:
+        lsoa_headroom: DataFrame with LSOA codes and distributed headroom.
+        households_df: DataFrame with LSOA codes and household counts.
+
+    Returns:
+        DataFrame with LSOA codes, distributed headroom, household counts, and headroom per household.
+    """
+    merged = pd.merge(lsoa_headroom, households_df, on="LSOA21CD", how="left")
+    merged["headroom_per_household"] = (
+        merged["distributed_headroom"] / merged["household_count"]
+    )
+    return merged[
+        [
+            "LSOA21CD",
+            "distributed_headroom",
+            "household_count",
+            "headroom_per_household",
+        ]
+    ]
+
+
+def process_substation_lsoa_data(
+    substations_df: gpd.GeoDataFrame,
+    lsoa_gdf: gpd.GeoDataFrame,
+    households_df: pd.DataFrame,
+) -> pd.DataFrame:
+    """
+    Process substation and LSOA data to calculate distributed headroom per household.
+
+    Args:
+        substations_df: GeoDataFrame of substations with headroom data.
+        lsoa_gdf: GeoDataFrame of LSOA boundaries.
+        households_df: DataFrame with household counts per LSOA.
+
+    Returns:
+        DataFrame with processed data including distributed headroom per household.
+    """
+    lsoa_headroom = distribute_substation_headroom(
+        substations_df, lsoa_gdf, households_df
+    )
+    return calculate_headroom_per_household(lsoa_headroom, households_df)
+
+
+def assess_heatpump_suitability(
```

Comment:
I don't have a good sense of capacity/voltage etc to feel like I can do a good job reviewing this function! @danlewis85 @crispy-wonton may be better placed

Reply:
I would suggest posting the numbers you've used in the team slack channel to get feedback on assumptions related to power for HPs :) they'll know best!